share/vss.go

```diff
@@ -0,0 +1,493 @@
+		return errDealAlreadyReceived
+
+	}
+	if a.deal == nil {
+		a.commits = d.Commitments
+		a.sid = d.SessionID
+		a.deal = d
+	}
+
+	if !validT(int(d.T), a.verifiers) {
+		return errors.New("verifier: invalid t received in Deal")
+	}
+
+	if !bytes.Equal(a.sid, d.SessionID) {
+		return errors.New("deal: sessionID is different from locally computed")
+	}
+
+	fi := d.SecShare
+	gi := d.RndShare
+	if fi.I != gi.I {
+		return errors.New("deal: not the same index for f and g share")
+	} else if fi.I < 0 || fi.I >= len(a.verifiers) {
+		return errors.New("deal: index out of bounds")
+	}
+	// compute fi * G + gi * H
+	fig := a.suite.Point().Base().Mul(nil, fi.V)
+	H := deriveH(a.suite, a.verifiers)
+	gih := a.suite.Point().Mul(H, gi.V)
+	ci := a.suite.Point().Add(fig, gih)
+
+	commitPoly := NewPubPoly(a.suite, nil, d.Commitments)
+
+	pubShare := commitPoly.Eval(fi.I)
+	if !ci.Equal(pubShare.V) {
+		return errors.New("deal: share do not verify against commitments")
+	}
+	return nil
+}
+
+func (a *aggregator) verifyComplaint(c *Complaint) error {
+	if err := a.verifyDeal(c.Deal, false); err == nil {
+		return errors.New("complaint: invalid because contains a valid deal")
+	}
+	if _, ok := a.complaints[c.Public.String()]; ok {
+		return errors.New("complaint: already stored one from same origin")
+	}
+	if !bytes.Equal(c.SessionID, a.sid) {
+		return errors.New("complaint: receiving inconsistent sessionID")
+	}
+
+	_, ok := findIndex(a.verifiers, c.Public)
+	if !ok {
+		return errors.New("complaint: from unknown participant")
+	}
+
+	if err := sign.VerifySchnorr(a.suite, c.Public, a.sid, c.Signature); err != nil {
+		return err
+	}
+
+	a.complaints[c.Public.String()] = c
+	return nil
+}
+
+func (a *aggregator) verifyApproval(ap *Approval) error {
+	if _, ok := a.approvals[ap.Public.String()]; ok {
+		return errors.New("approval: already stored one from same origin")
+	}
+
+	if !bytes.Equal(ap.SessionID, a.sid) {
+		return errors.New("approval: does not match session id recorded")
+	}
+
+	if err := sign.VerifySchnorr(a.suite, ap.Public, a.sid, ap.Signature); err != nil {
+		return err
+	}
+
+	a.approvals[ap.Public.String()] = ap
+	return nil
+}
+
+func (a *aggregator) EnoughApprovals() bool {
+	if a.deal == nil {
+		return false
+	}
+	return len(a.approvals) >= a.t
+}
+
+func (a *aggregator) DealCertified() bool {
+	if len(a.complaints) >= a.t-1 {
+		return false
+	}
+	return true
+}
+
+func minimumT(verifiers []abstract.Point) int {
+	return (len(verifiers) + 1) / 2
+}
+
+func validT(t int, verifiers []abstract.Point) bool {
+	return t >= minimumT(verifiers) && t <= len(verifiers) && int(uint32(t)) == t
+}
+
+// HashFunc is used to compute
+//  - the second base H out of a list of public keys
+//  - the hash of the message to sign in a Complaint
+var HashFunc = sha256.New
+
+func deriveH(suite abstract.Suite, verifiers []abstract.Point) abstract.Point {
+	var b bytes.Buffer
+	for i := range verifiers {
+		verifiers[i].MarshalTo(&b)
+	}
+	h := HashFunc()
```

Comment:
I agree that with `suite.Hash()` we do not know which hash function is used. However, in most cases it will be `sha256` and somebody reading the code might wonder at this point "Huh, why is it important here to explicitly use SHA256?!". I think for consistency it would be better to just use `suite.Hash()`, yes.